### DISCUSSION TITLE
Render of a context causes whole app to reset state

### DISCUSSION THREAD

**Discussion by incammicia** [vercel/next.js#41955] [Help] [answered]
My _app.tsx return looks like this

```
<AuthProvider>
  <ApiProvider>
    <Header />
    <Component {...pageProps} />
  </ApiProvider>
</AuthProvider>
```
with a bunch of other components and providers. The app mostly uses client-side api calls with React Query.
AuthProvider has a query that runs every 5 minutes to refresh the access token.
When the token is refreshed, ApiProvider gets updated (every api call needs the new token).
So, it makes sense that the whole app renders, because every page uses the contexts and that's no big deal.

What's weird is that when it happens, every state of the page gets resetted (this happens also in Header).
If I have a modal open (with a simple state from useState), it closes.
This is similar to unmounting and mounting a component back and I've already tried a few solutions like memoizing and such.

This is what I expect to happen:
https://codesandbox.io/s/prod-shadow-6841rk
The context gets updated, the subscriber gets rendered, but its state doesn't get resetted

Does anyone know why this doesn't happen in my application? Is there something about Next that I'm missing?

PS. The codebase is quite huge, that's why I didn't include code, maybe someone already has an idea of the problem just by reading the info above, but if that's not the case, I can prepare a git repo with a demo.

Thank you in advance

**Comment by icyJoseph** [comment, marked as answer]
I have seen this before with auth providers that when they enter a loading phase, return null or some spinner, instead of children, which of course means everything under the tree gets unmounted. 

Many months ago I saw this and the solution was to pass a flag to the auth provider, which in that case came from `next-auth` IIRC.

**Reply by incammicia** [reply]
@icyJoseph thank you so much, your answer gave me the correct insight to solve it (and it was simpler that I thought).
I have both a custom AuthProvider and ApiProvider and I had indeed a "if (!data) return null" where data was returned by react-query.
I noticed that just after that token refresh, react-query returned "undefined" for an instant, before getting the new data (invisible to human eye).
The solution was very simple, pass "keepPreviousData: true" to the options of react-query, this way I still have the previous API Client in the meantime that the token gets refreshed. And this behaviour makes a lot of sense.
Thank you again!
